File: data_structure/map/src/MapDemo.cs
// C#
// データ構造: マップ (Map)

using System;
using System.Collections.Generic;
using System.Linq;

class MapData
{
    private Dictionary<string, int> _data;

    public MapData()
    {
        _data = new Dictionary<string, int>();
    }

    public List<KeyValuePair<string, int>> Get()
    {
        return _data.ToList();
    }
    
    public List<string> GetKeys()
    {
        return _data.Keys.ToList();
    }
    
    public List<int> GetValues()
    {
        return _data.Values.ToList();
    }
    
    public string GetKey(int value)
    {
        try
        {
            return _data.FirstOrDefault(x => x.Value == value).Key;
        }
        catch
        {
            Console.WriteLine($"ERROR: {value} は範囲外です");
            return null;
        }
    }
    
    public int? GetValue(string key)
    {
        if (_data.TryGetValue(key, out int value))
        {
            return value;
        }
        
        Console.WriteLine($"ERROR: {key} は範囲外です");
        return null;
    }

    public bool Add(string key, int value)
    {
        if (_data.ContainsKey(key))
        {
            Console.WriteLine($"ERROR: {key} は重複です");
            return false;
        }
        
        _data[key] = value;
        return true;
    }
    
    public bool Remove(string key)
    {
        if (_data.Remove(key))
        {
            return true;
        }
        
        Console.WriteLine($"ERROR: {key} は範囲外です");
        return false;
    }
    
    public bool Update(string key, int value)
    {
        if (_data.ContainsKey(key))
        {
            _data[key] = value;
            return true;
        }
        
        Console.WriteLine($"ERROR: {key} は範囲外です");
        return false;
    }
    
    public bool IsEmpty()
    {
        return _data.Count == 0;
    }
    
    public int Size()
    {
        return _data.Count;
    }
    
    public bool Clear()
    {
        _data.Clear();
        return true;
    }

    static void Main(string[] args)
    {
        Console.WriteLine("Map TEST -----> start");

        Console.WriteLine("\nnew");
        MapData mapData = new MapData();
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nis_empty");
        bool output = mapData.IsEmpty();
        Console.WriteLine($"  出力値: {output}");

        Console.WriteLine("\nsize");
        int size = mapData.Size();
        Console.WriteLine($"  出力値: {size}");

        Console.WriteLine("\nadd");
        (string, int) input = ("apple", 100);
        Console.WriteLine($"  入力値: {input}");
        bool addOutput = mapData.Add(input.Item1, input.Item2);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nadd");
        input = ("banana", 150);
        Console.WriteLine($"  入力値: {input}");
        addOutput = mapData.Add(input.Item1, input.Item2);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nadd");
        input = ("apple", 200);
        Console.WriteLine($"  入力値: {input}");
        addOutput = mapData.Add(input.Item1, input.Item2);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nsize");
        size = mapData.Size();
        Console.WriteLine($"  出力値: {size}");

        Console.WriteLine("\nget");
        string getInput = "apple";
        Console.WriteLine($"  入力値: {getInput}");
        int? getOutput = mapData.GetValue(getInput);
        Console.WriteLine($"  出力値: {getOutput}");

        Console.WriteLine("\nget");
        getInput = "orange";
        Console.WriteLine($"  入力値: {getInput}");
        getOutput = mapData.GetValue(getInput);
        Console.WriteLine($"  出力値: {getOutput}");

        Console.WriteLine("\nupdate");
        input = ("banana", 180);
        Console.WriteLine($"  入力値: {input}");
        bool updateOutput = mapData.Update(input.Item1, input.Item2);
        Console.WriteLine($"  出力値: {updateOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nupdate");
        input = ("orange", 250);
        Console.WriteLine($"  入力値: {input}");
        updateOutput = mapData.Update(input.Item1, input.Item2);
        Console.WriteLine($"  出力値: {updateOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nget");
        getInput = "banana";
        getOutput = mapData.GetValue(getInput);
        Console.WriteLine($"  出力値: {getOutput}");

        Console.WriteLine("\nget_keys");
        var keysOutput = mapData.GetKeys();
        Console.WriteLine($"  出力値: {string.Join(", ", keysOutput)}");

        Console.WriteLine("\nvalues");
        var valuesOutput = mapData.GetValues();
        Console.WriteLine($"  出力値: {string.Join(", ", valuesOutput)}");

        Console.WriteLine("\nget_key");
        int keyInput = 180;
        Console.WriteLine($"  入力値: {keyInput}");
        string keyOutput = mapData.GetKey(keyInput);
        Console.WriteLine($"  出力値: {keyOutput}");

        Console.WriteLine("\nget_key");
        keyInput = 500;
        Console.WriteLine($"  入力値: {keyInput}");
        keyOutput = mapData.GetKey(keyInput);
        Console.WriteLine($"  出力値: {keyOutput}");

        Console.WriteLine("\nremove");
        string removeInput = "apple";
        Console.WriteLine($"  入力値: {removeInput}");
        bool removeOutput = mapData.Remove(removeInput);
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nremove");
        removeInput = "orange";
        Console.WriteLine($"  入力値: {removeInput}");
        removeOutput = mapData.Remove(removeInput);
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nsize");
        size = mapData.Size();
        Console.WriteLine($"  出力値: {size}");

        Console.WriteLine("\nget_keys");
        keysOutput = mapData.GetKeys();
        Console.WriteLine($"  出力値: {string.Join(", ", keysOutput)}");

        Console.WriteLine("\nclear");
        bool clearOutput = mapData.Clear();
        Console.WriteLine($"  出力値: {clearOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", mapData.Get())}");

        Console.WriteLine("\nsize");
        size = mapData.Size();
        Console.WriteLine($"  出力値: {size}");

        Console.WriteLine("\nis_empty");
        output = mapData.IsEmpty();
        Console.WriteLine($"  出力値: {output}");

        Console.WriteLine("\nMap TEST <----- end");
    }
}

File: data_structure/map/src/MapDemo.java
// Java
// データ構造: マップ (Map)

import java.util.HashMap;
import java.util.Map;

class MapData {
    private Map<String, Integer> data;

    public MapData() {
        data = new HashMap<>();
    }

    public Map<String, Integer> get() {
        return new HashMap<>(data);
    }

    public java.util.Set<String> getKeys() {
        return data.keySet();
    }

    public java.util.Collection<Integer> getValues() {
        return data.values();
    }

    public String getKey(Integer value) {
        for (Map.Entry<String, Integer> entry : data.entrySet()) {
            if (entry.getValue().equals(value)) {
                return entry.getKey();
            }
        }
        System.out.printf("ERROR: %d は範囲外です%n", value);
        return null;
    }

    public Integer getValue(String key) {
        if (data.containsKey(key)) {
            return data.get(key);
        }
        System.out.printf("ERROR: %s は範囲外です%n", key);
        return null;
    }

    public boolean add(String key, Integer value) {
        if (data.containsKey(key)) {
            System.out.printf("ERROR: %s は重複です%n", key);
            return false;
        }
        data.put(key, value);
        return true;
    }

    public boolean remove(String key) {
        if (data.containsKey(key)) {
            data.remove(key);
            return true;
        }
        System.out.printf("ERROR: %s は範囲外です%n", key);
        return false;
    }

    public boolean update(String key, Integer value) {
        if (data.containsKey(key)) {
            data.put(key, value);
            return true;
        }
        System.out.printf("ERROR: %s は範囲外です%n", key);
        return false;
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    public int size() {
        return data.size();
    }

    public boolean clear() {
        data.clear();
        return true;
    }

    public static void main(String[] args) {
        System.out.println("Map TEST -----> start");

        System.out.println("\nnew");
        MapData mapData = new MapData();
        System.out.printf("  現在のデータ: %s%n", mapData.get());

        System.out.println("\nis_empty");
        boolean output = mapData.isEmpty();
        System.out.printf("  出力値: %b%n", output);

        System.out.println("\nsize");
        int sizeOutput = mapData.size();
        System.out.printf("  出力値: %d%n", sizeOutput);

        System.out.println("\nadd");
        String[] inputs = {"apple", "banana", "apple"};
        int[] values = {100, 150, 200};
        for (int i = 0; i < inputs.length; i++) {
            System.out.printf("  入力値: (%s, %d)%n", inputs[i], values[i]);
            boolean addOutput = mapData.add(inputs[i], values[i]);
            System.out.printf("  出力値: %b%n", addOutput);
            System.out.printf("  現在のデータ: %s%n", mapData.get());
        }

        System.out.println("\nsize");
        sizeOutput = mapData.size();
        System.out.printf("  出力値: %d%n", sizeOutput);

        System.out.println("\nget");
        String[] getInputs = {"apple", "orange"};
        for (String input : getInputs) {
            System.out.printf("  入力値: %s%n", input);
            Integer getOutput = mapData.getValue(input);
            System.out.printf("  出力値: %s%n", getOutput);
        }

        System.out.println("\nupdate");
        String[] updateKeys = {"banana", "orange"};
        int[] updateValues = {180, 250};
        for (int i = 0; i < updateKeys.length; i++) {
            System.out.printf("  入力値: (%s, %d)%n", updateKeys[i], updateValues[i]);
            boolean updateOutput = mapData.update(updateKeys[i], updateValues[i]);
            System.out.printf("  出力値: %b%n", updateOutput);
            System.out.printf("  現在のデータ: %s%n", mapData.get());
        }

        System.out.println("\nget");
        String getValue = "banana";
        Integer getOutput = mapData.getValue(getValue);
        System.out.printf("  出力値: %s%n", getOutput);

        System.out.println("\nget_keys");
        java.util.Set<String> keysOutput = mapData.getKeys();
        System.out.printf("  出力値: %s%n", keysOutput);

        System.out.println("\nvalues");
        java.util.Collection<Integer> valuesOutput = mapData.getValues();
        System.out.printf("  出力値: %s%n", valuesOutput);

        System.out.println("\nget_key");
        int[] keyInputs = {180, 500};
        for (int input : keyInputs) {
            System.out.printf("  入力値: %d%n", input);
            String keyOutput = mapData.getKey(input);
            System.out.printf("  出力値: %s%n", keyOutput);
        }

        System.out.println("\nremove");
        String[] removeInputs = {"apple", "orange"};
        for (String input : removeInputs) {
            System.out.printf("  入力値: %s%n", input);
            boolean removeOutput = mapData.remove(input);
            System.out.printf("  出力値: %b%n", removeOutput);
            System.out.printf("  現在のデータ: %s%n", mapData.get());
        }

        System.out.println("\nsize");
        sizeOutput = mapData.size();
        System.out.printf("  出力値: %d%n", sizeOutput);

        System.out.println("\nget_keys");
        keysOutput = mapData.getKeys();
        System.out.printf("  出力値: %s%n", keysOutput);

        System.out.println("\nclear");
        boolean clearOutput = mapData.clear();
        System.out.printf("  出力値: %b%n", clearOutput);
        System.out.printf("  現在のデータ: %s%n", mapData.get());

        System.out.println("\nsize");
        sizeOutput = mapData.size();
        System.out.printf("  出力値: %d%n", sizeOutput);

        System.out.println("\nis_empty");
        output = mapData.isEmpty();
        System.out.printf("  出力値: %b%n", output);

        System.out.println("\nMap TEST <----- end");
    }
}